### PR TITLE
Support instanceof and checkcast for array types (attempt 2, using a separate field)

### PR DESCRIPTION
**Pull request by smowton**
Similar to https://github.com/diffblue/cbmc/pull/5033, except instead of encoding array types into the class identifier (e.g. `java::array[reference(MyType)]`, we store them in a separate fields named `@array_dimensions` and `@element_class_identifier`. For example, type `MyType[][]` is represented by `@class_identifier` == `java::array[reference]`, `@array_dimensions` == `2` and `@element_class_identifier` == `java::MyType`.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e77ed8ce092d6073cdebc827882a11ced536f7ea).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/125060196

File: jbmc/src/java_bytecode/remove_instanceof.cpp [hunk 1]
```diff
@@ -56,6 +56,41 @@ class remove_instanceoft
...
+  children.push_back(target_type);
+  // Sort alphabetically to make order of generated disjuncts
+  // independent of class loading order
+  std::sort(
+    children.begin(), children.end(), [](const irep_idt &a, const irep_idt &b) {
+      return a.compare(b) < 0;
+    });
+
+  exprt::operandst or_ops;
+  for(const auto &clsname : children)
```
Comment (thk123) [inline jbmc/src/java_bytecode/remove_instanceof.cpp]: :pick: `class_name` (I always have to squint at this abbreviation)
:pick: this could be done using a `std::transform` to make it clearer the whole vector is being covered.

File: jbmc/src/java_bytecode/remove_instanceof.cpp [hunk 2]
```diff
@@ -56,6 +56,41 @@ class remove_instanceoft
...
+  {
+    constant_exprt clsexpr(clsname, string_typet());
```
Comment (thk123) [inline jbmc/src/java_bytecode/remove_instanceof.cpp]: :pick: `class_expr`

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp [hunk 3]
```diff
@@ -829,6 +829,26 @@ void add_java_array_types(symbol_tablet &symbol_table)
     data_component.set_base_name("data");
     class_type.components().push_back(data_component);
 
+    if(l == 'a')
+    {
+      // This is a reference array (java::array[reference]). Add extra fields to
+      // carry the innermost element type and array dimension.
+      java_class_typet::componentt array_element_classid_component(
+        JAVA_ARRAY_ELEMENT_CLASSID_FIELD_NAME, string_typet());
```
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:837]: :bulb: pull out a helper function for making a component with same pretty/base/name

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp [hunk 4]
```diff
@@ -888,6 +908,26 @@ void add_java_array_types(symbol_tablet &symbol_table)
     java_new_array.copy_to_operands(old_length);
     code_assignt create_blank(local_symexpr, java_new_array);
 
+    codet copy_type_information = code_skipt();
+    if(l == 'a')
+    {
+      // Reference arrays carry additional type information in their classids
+      // which should be copied:
+      member_exprt old_array_dimension(
```
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.cpp]: :bulb: Rather than re-specifying the types of these components, it would be better to build the `member_exprt` from the componentt of the type (which I think you can do?)

Comment (thk123) [reply jbmc/src/java_bytecode/java_bytecode_convert_class.cpp]: Or (re)use `get_array_dimension_field`

File: jbmc/src/java_bytecode/remove_instanceof.cpp [hunk 5]
```diff
@@ -132,17 +134,34 @@ bool remove_instanceoft::lower_instanceof(
...
 
-  // Replace the instanceof operation with (check_ptr != null &&
-  //  (check_ptr->@class_identifier == "A" ||
-  //   check_ptr->@class_identifier == "B" || ...
+  const irep_idt &underlying_name =
+    underlying_type_and_dimension.first.get_identifier();
 
-  // By operating directly on a pointer rather than using a temporary
+  auto jlo = to_struct_tag_type(java_lang_object_type().subtype());
+  bool underlying_type_is_object = underlying_type_and_dimension.first == jlo;
```
Comment (thk123) [inline jbmc/src/java_bytecode/remove_instanceof.cpp]: So `Foo[]` is *not* a super type of `Foo[][]`?

File: jbmc/src/java_bytecode/remove_instanceof.cpp [hunk 6]
```diff
@@ -151,13 +170,49 @@ bool remove_instanceoft::lower_instanceof(
...
+  else if(!underlying_type_is_object)
+  {
+    test_conjuncts.push_back(subtype_expr(
+      get_class_identifier_field(check_ptr, jlo, ns),
+      underlying_name,
+      class_hierarchy));
+  }
 
-  expr = and_exprt(not_null, subtype_check);
+  expr = conjunction(test_conjuncts);
```
Comment (thk123) [inline jbmc/src/java_bytecode/remove_instanceof.cpp:216]: :bulb: This logic seems like a great candidate for a unit test (that hopefully `simplify_expr` with constant strings will allow you to actually test the outcome).

File: jbmc/src/java_bytecode/remove_java_new.cpp [hunk 7]
```diff
@@ -153,22 +153,52 @@ goto_programt::targett remove_java_newt::lower_java_new_array(
 
   const struct_typet &struct_type = to_struct_type(ns.follow(object_type));
 
-  // Ideally we would have a check for `is_valid_java_array(struct_type)` but
-  // `is_valid_java_array is part of the java_bytecode module and we cannot
-  // introduce such dependencies. We do this simple check instead:
-  PRECONDITION(struct_type.components().size() == 3);
+  PRECONDITION(is_valid_java_array(struct_type));
```
Comment (thk123) [inline jbmc/src/java_bytecode/remove_java_new.cpp:156]: :pick: should have gone in separate commit

File: jbmc/src/java_bytecode/remove_java_new.cpp [hunk 8]
```diff
@@ -153,22 +153,52 @@ goto_programt::targett remove_java_newt::lower_java_new_array(
...
+  // fields (if the given type is void then the caller must set these up itself;
+  // this happens in array[reference].clone(), where the type info must be
+  // copied over from the input array)
+  if(
+    java_array_element_type(object_type) !=
+    java_reference_type(java_void_type()))
+  {
+    const auto underlying_type_and_dimension =
+      java_array_dimension_and_element_type(object_type);
+    if(underlying_type_and_dimension.second >= 1)
```
Comment (thk123) [inline jbmc/src/java_bytecode/remove_java_new.cpp]: :question: should this be an invariant? if we have an array type, shouldn't the dimension always be `> 0`?

**Review by thk123 — COMMENTED** [review]
Nothing blocking - but one question to check I understand the code. Otherwise LGTM

File: jbmc/src/java_bytecode/remove_instanceof.cpp (hunk 5 above)
Comment (smowton) [reply jbmc/src/java_bytecode/remove_instanceof.cpp]: It is not, much like `Foo` and `Foo[]` are unrelated types. Only `Object[][]` and `Object[]` are related this way, by virtue of `Object[]` being an `Object`. `Foo[][]` *is* an `Object[]` therefore.

File: jbmc/src/java_bytecode/remove_java_new.cpp (hunk 8 above)
Comment (smowton) [reply jbmc/src/java_bytecode/remove_java_new.cpp]: We also get 0 for primitive arrays; renamed this `java_*reference*_array_....` accordingly

Comment (thk123) [reply jbmc/src/java_bytecode/remove_java_new.cpp]: That's surprising... doesn't that mean `assert int[] instance Object[]` fails? Yes: as expected in the JVM (but not expected by me).

Comment (smowton) [reply jbmc/src/java_bytecode/remove_java_new.cpp]: yep, the primitive arrays are their own thing, subtypes of Object but unrelated to Object[]

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp [hunk 9]
```diff
@@ -886,6 +908,39 @@ void add_java_array_types(symbol_tablet &symbol_table)
...
+      // Reference arrays carry additional type information in their classids
+      // which should be copied:
+      const auto &array_dimension_component =
+        class_type.get_component(JAVA_ARRAY_DIMENSION_FIELD_NAME);
+      const auto &array_element_classid_component =
+        class_type.get_component(JAVA_ARRAY_ELEMENT_CLASSID_FIELD_NAME);
+
+      member_exprt old_array_dimension(
+        old_array,
+        array_dimension_component.get_name(),
```
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.cpp]: :bulb: Sorry should have been more explicit - this is definitely better, but you can even go one further:
```suggestion
        member_exprt old_array_dimension(old_array, array_dimension_component)
```
using the constructor that takes a `struct_typet::componentt` directly

File: jbmc/src/java_bytecode/remove_instanceof.cpp [hunk 10]
```diff
@@ -94,30 +131,38 @@ bool remove_instanceoft::lower_instanceof(
...
-  // Replace the instanceof operation with (check_ptr != null &&
-  //  (check_ptr->@class_identifier == "A" ||
-  //   check_ptr->@class_identifier == "B" || ...
+  const irep_idt &underlying_name =
+    to_struct_tag_type(underlying_type_and_dimension.first).get_identifier();
 
-  // By operating directly on a pointer rather than using a temporary
+  auto jlo = to_struct_tag_type(java_lang_object_type().subtype());
+  bool underlying_type_is_object = underlying_type_and_dimension.first == jlo;
+  bool target_type_is_array = underlying_type_and_dimension.second > 0;
```
Comment (thk123) [inline jbmc/src/java_bytecode/remove_instanceof.cpp]: :no_entry_sign: Perhaps re-reviewing too soon - but this variable should definitely be called `target_type_is_reference_array` given the dimension is zero for primitive arrays

File: jbmc/src/java_bytecode/remove_java_new.cpp [hunk 11]
```diff
@@ -153,22 +153,50 @@ goto_programt::targett remove_java_newt::lower_java_new_array(
...
 
+  // If it's a reference array we need to set the dimension and element type
+  // fields (if the given type is void then the caller must set these up itself;
+  // this happens in array[reference].clone(), where the type info must be
+  // copied over from the input array)
+  const auto underlying_type_and_dimension =
+    java_reference_array_dimension_and_element_type(object_type);
+
+  if(
+    underlying_type_and_dimension.second >= 1 &&
```
Comment (thk123) [inline jbmc/src/java_bytecode/remove_java_new.cpp:177]: :pick: might be even clearer to introduce a tempoary

```
const bool is_reference_array  = underlying_type_and_dimension.second >= 1
```

**Review by thk123 — COMMENTED** [review]
Sorry one final blocking comment - just feel like `dimension(int[]) == 0` is quite a gotcha so need to make variable names as clear as possible (alternative suggestion: what would break if these also had dimension one, and the `instanceof Object[]` special cased when lhs is primitive array?

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 660e309c77e271bf7d2657261ba20f5867c426f3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/125237846
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Comment by smowton** [comment]
@thk123 all done

File: jbmc/src/java_bytecode/remove_java_new.cpp (hunk 11 above)
Comment (smowton) [reply jbmc/src/java_bytecode/remove_java_new.cpp:177]: That's actually not accurate, as an array-of-`void*` passes that test, as per the comment above

File: jbmc/src/java_bytecode/java_types.cpp [hunk 12]
```diff
@@ -182,6 +182,66 @@ bool is_multidim_java_array_type(const typet &type)
...
+static typet get_array_element_subtype(const typet &type)
+{
+  const typet &array_type = to_java_reference_type(type).subtype();
+  return java_array_element_type(to_struct_tag_type(array_type));
+}
+
+/// Returns the underlying element type and array dimensionality of Java struct
+/// \p type. If this is not an array type the return value will be `{type, 0}`.
+std::pair<typet, std::size_t>
+java_array_dimension_and_element_type(const struct_tag_typet &type)
```
Comment (thk123) [inline jbmc/src/java_bytecode/java_types.cpp:188]: :no_entry_sign: out of band discussion - this function has changed name, but still only returns non-zero dimension for *reference* type arrays. Either the name should remain `java_reference_array_dimension...` or it should handle primitive array dimensions

**Review by thk123 — APPROVED** [review]
Thanks

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 3630ecfc4fd861a6e828621634b708c63b4d2f68).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/125247670
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 6473565816c823177d48478184355463f066c4b4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/125251880

File: jbmc/src/java_bytecode/java_types.cpp [hunk 13]
```diff
@@ -182,6 +182,24 @@ bool is_multidim_java_array_type(const typet &type)
                                        to_struct_tag_type(type.subtype())));
 }
 
+/// Returns the underlying element type and array dimensionality of Java struct
+/// \p type. If this is not an array type the return value will be `{type, 0}`.
+std::pair<typet, std::size_t>
+java_array_dimension_and_element_type(const struct_tag_typet &type)
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_types.cpp:188]: Add into the commit message that this is bumping java-models-library, or even better move the bump into a separate commit.

**Review by peterschrammel — APPROVED** [review]
Seems to work.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 9440702bc136353ac27b96343625e16fef3ee311).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/125315391
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9a6361e2a11399db1ab07836277f51d8f40f2bf7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/125315840
